samples/opengl/depthMapTriangulation/decimate.cpp

```diff
@@ -0,0 +1,680 @@
+
+    MyMesh::FaceVertexCWIter fv_it = current_mesh->cfv_cwbegin(fh);
+    MyMesh::VertexHandle vh0       = *fv_it;
+    MyMesh::VertexHandle vh1       = *(++fv_it);
+    MyMesh::VertexHandle vh2       = *(++fv_it);
+    OpenMesh::Vec3f v0             = current_mesh->point(vh0);
+    OpenMesh::Vec3f v1             = current_mesh->point(vh1);
+    OpenMesh::Vec3f v2             = current_mesh->point(vh2);
+
+    OpenMesh::Vec3f normal = OpenMesh::cross((v1 - v0), (v2 - v0));
+
+    float area = normal.norm();
+    normal /= area;
+    area *= 0.5;
+
+    float d = -(normal[0] * v0[0]) - (normal[1] * v0[1]) - (normal[2] * v0[2]);
+
+    mat4 k_p;
+    k_p(0, 0) = normal[0] * normal[0] * area;
+    k_p(0, 1) = normal[0] * normal[1] * area;
+    k_p(0, 2) = normal[0] * normal[2] * area;
+    k_p(0, 3) = normal[0] * d * area;
+
+    k_p(1, 0) = k_p(0, 1);
+    k_p(1, 1) = normal[1] * normal[1] * area;
+    k_p(1, 2) = normal[1] * normal[2] * area;
+    k_p(1, 3) = normal[1] * d * area;
+
+    k_p(2, 0) = k_p(0, 2);
+    k_p(2, 1) = k_p(1, 2);
+    k_p(2, 2) = normal[2] * normal[2] * area;
+    k_p(2, 3) = normal[2] * d * area;
+
+    k_p(3, 0) = k_p(0, 3);
+    k_p(3, 1) = k_p(1, 3);
+    k_p(3, 2) = k_p(2, 3);
+    k_p(3, 3) = d * d * area;
+
+    return k_p;
+}
+
+bool QuadricDecimater::check_for_folding_triangles(const MyMesh::EdgeHandle edge)
+{
+    // check if any of the triangles that are altered by this collapse have normals with greater
+    // angles than 60° to each other
+    // --> loop through all faces of the vertices of the cendidat edge and check their normals against each other
+
+    // collect all normals
+    std::vector<OpenMesh::Vec3f> normals;
+    MyMesh::VertexFaceCWIter fcw_it =
+        current_mesh->cvf_cwbegin(current_mesh->to_vertex_handle(current_mesh->halfedge_handle(edge, 0)));
+    MyMesh::VertexFaceCWIter fcw_it2 =
+        current_mesh->cvf_cwbegin(current_mesh->from_vertex_handle(current_mesh->halfedge_handle(edge, 0)));
+
+    while (fcw_it.is_valid())
+    {
+        if (current_mesh->status(*fcw_it).deleted()) continue;
+        normals.push_back(current_mesh->normal(*fcw_it));
+        ++fcw_it;
+    }
+    while (fcw_it2.is_valid())
+    {
+        if (current_mesh->status(*fcw_it2).deleted()) continue;
+        normals.push_back(current_mesh->normal(*fcw_it2));
+        ++fcw_it2;
+    }
+
+    // check them against each other
+    for (int i = 0; i < normals.size(); ++i)
+    {
+        OpenMesh::Vec3f normal_1 = normals[i];
+
+        for (int j = i + 1; j < normals.size(); ++j)
+        {
+            // dot(A,B) = |A| * |B| * cos(angle)
+            // which can be rearranged to
+            // angle = arccos(dot(A, B) / (|A| * |B|)).
+            float angle = OpenMesh::dot(normal_1, normals[j]);
+            angle       = acos(angle);
+
+            // pi<float>() / 3.0f rad = 60 degree
+            // pi<float>() / 3.0f = 1.0471975512f
+            if (angle > 1.0471975512f)
```

Comment:
just write `radians(60.f)`. The compiler will compute this at compile time and it's easier readable.